Open Judge System/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs

```diff
@@ -0,0 +1,119 @@
+﻿namespace OJS.Services.Data.ParticipantScores
+{
+    using System.Linq;
+    using System.Transactions;
+
+    using EntityFramework.Extensions;
+
+    using OJS.Data.Models;
+    using OJS.Data.Repositories.Contracts;
+    using OJS.Services.Data.Participants;
+
+    public class ParticipantScoresDataService : IParticipantScoresDataService
+    {
+        private readonly IEfGenericRepository<ParticipantScore> participantScores;
+        private readonly IParticipantsDataService participantsData;
+
+        public ParticipantScoresDataService(
+            IEfGenericRepository<ParticipantScore> participantScores,
+            IParticipantsDataService participantsData)
+        {
+            this.participantScores = participantScores;
+            this.participantsData = participantsData;
+        }
+
+        public ParticipantScore Get(int participantId, int problemId) =>
+            this.participantScores
+                .All()
+                .FirstOrDefault(ps => ps.ParticipantId == participantId &&
+                    ps.ProblemId == problemId);
+
+        public ParticipantScore Get(int participantId, int problemId, bool isOfficial) =>
+            this.participantScores
+                .All()
+                .FirstOrDefault(ps => ps.ParticipantId == participantId &&
+                    ps.ProblemId == problemId &&
+                    ps.IsOfficial == isOfficial);
+
+        public IQueryable<ParticipantScore> GetAll()
+            => this.participantScores.All();
+
+        public bool Save(Submission submission, bool resetScore = false)
+        {
+            using (var transaction = new TransactionScope())
+            {
+                if (submission.ParticipantId == null || submission.ProblemId == null)
+                {
+                    return false;
+                }
+
+                var participant = this.participantsData
+                        .GetByIdQuery(submission.ParticipantId.Value)
+                        .Select(p => new
+                        {
+                            p.IsOfficial,
+                            p.User.UserName
+                        })
+                        .FirstOrDefault();
+
+                if (participant == null)
+                {
+                    return false;
```

Comment:
It's not a problem at all, since the transaction scope has nothing to "Complete" on this stage of the execution, and the disposal of the scope is guaranteed anyway by the `using{}`